### DISCUSSION TITLE
Question: Checkstyle tooling API vs Parsing output

### DISCUSSION THREAD

**Discussion by jdneo** [jdneo/vscode-checkstyle#355] [Q&A] [answered]
Originally posted at https://github.com/jdneo/vscode-checkstyle/issues/353

This isn't an issue, just a curiosity I have.

Looking at the implementation of this plugin, it looks pretty advanced, interacting with the Checkstyle tooling API and language server etc.

Looking at one of the other plugins, e.g. https://github.com/thomdabeast/java-checkstyle-vscode
It seems to be implemented in a much simpler, perhaps more naive, way. It basically just executes the checkstyle jar with `java -jar` and a bunch of arguments, and then parses the output to generate warnings.

Is there an inherent disadvantage to this perhaps more naive approach?

EDIT:
An advantage of the naive approach that I can see is that things like the suppressions functionality (as requested here https://github.com/jdneo/vscode-checkstyle/issues/335) would work out of the box I think.

**Comment by jdneo** [comment, marked as answer]
@sbreakey

Actually, for a long time ago, this extension also used the naive approach, call the command line and parse the output. But this approach has some problems, especially the cost of providing real-time check is high because for each check, a jvm will be launched.

That's why I changed the implementation to the current way, using the api. Meanwhile, using the Java code makes it possible to provide quick fixes for the violations (parsing the AST and change it)

> An advantage of the naive approach that I can see is that things like the suppressions functionality.

I haven't investigated this area yet, maybe the tooling api has such capability?

**Comment by sbreakey** [comment]
@jdneo thanks for the explanation. That makes a lot of sense, I didn't think about the jvm cost.
Not to mention the added benefit of being able to provide quick fixes!

> I haven't investigated this area yet, maybe the tooling api has such capability?

Chat GPT says:
> As of my last update in September 2021, Checkstyle's API did not provide a direct, comprehensive mechanism for handling suppressions programmatically.

:(

**Comment by jdneo** [comment]
I think the recommended way to support suppression filter from Checkstyle official is to write it into Checkstyle configuration files: https://checkstyle.sourceforge.io/filters/suppressionfilter.html

By saying the suppression support, do you mean things like https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/suppressions-filter.html

**Comment by sbreakey** [comment]
> I think the recommended way to support suppression filter from Checkstyle official is to write it into Checkstyle configuration files: https://checkstyle.sourceforge.io/filters/suppressionfilter.html

Yes, exactly. IMO it should be supported via the `SuppressionFilter` and not via configuration as suggested in https://github.com/jdneo/vscode-checkstyle/issues/335

Meaning, I don't think adding configuration to this extension like `java.checkstyle.suppressionsLocation` or `java.checkstyle.ignoreViolations` makes sense. It seems to just duplicate functionality that already exists in checkstyle itself? The plugin just needs to somehow hook into that functionality.

Even the google and sun checks already include a SuppressionFilter, e.g. https://github.com/checkstyle/checkstyle/blob/7d53438504c20b46e025c0124a63d3cdc0b9fde5/src/main/resources/google_checks.xml#L33C1-L37C12
```xml
  <module name="SuppressionFilter">
    <property name="file" value="${org.checkstyle.google.suppressionfilter.config}"
           default="checkstyle-suppressions.xml" />
    <property name="optional" value="true"/>
  </module>
```

So if someone is using sun or google checks, they can still specify their own suppression file.

> By saying the suppression support, do you mean things like https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/suppressions-filter.html

Yes, but as I mentioned above, IMO, this should be controlled through the SuppressionFilter in a checkstyle XML and NOT by config properties in the extension (or in a gradle/maven plugin).

**Reply by jdneo** [reply]
Yes, I think we are on the same page :)
